Adding editing of packages on github

mmoskal opened the pull request:
This PR adds support for:
* import of existing packages from github for editing
* pushing changes of imported github packages
* upon commit (integrated with push/pull) users can specify check in message and bump
* we use github for merging
* in case automatic merging fails, a PR is created on github and user is moved to current master
* listing all repos in user's account and importing it
* creating new repos
* initializing existing repos (adding `pxt.json` and friends)
* adding packages in local workspace as extensions
* logging in via github - the token is obtained through our cloud, but is only stored locally; all github operations except for login are done without involving our cloud

This should dramatically improve user's ability to author packages. I have so far only tested this with localhost, but should work on cloud instances. Waiting on backend PR for this.

TODO:
* [x] replace `workspace:...` deps with `github:...` upon publishing

<img width="771" alt="screen shot 2018-07-17 at 11 58 32" src="https://user-images.githubusercontent.com/10673976/42813463-04223a36-89b9-11e8-8c24-b94dff7390bf.png">
<img width="760" alt="screen shot 2018-07-17 at 11 58 39" src="https://user-images.githubusercontent.com/10673976/42813471-0971e374-89b9-11e8-882c-aee6348e008e.png">
<img width="812" alt="screen shot 2018-07-17 at 11 59 00" src="https://user-images.githubusercontent.com/10673976/42813476-0b9daa8e-89b9-11e8-8534-6e3b5a0bce84.png">

<img width="752" alt="screen shot 2018-07-16 at 14 06 36" src="https://user-images.githubusercontent.com/10673976/42758007-4709bb74-8902-11e8-961e-efa58e9da7eb.png">
<img width="770" alt="screen shot 2018-07-16 at 14 06 46" src="https://user-images.githubusercontent.com/10673976/42758013-4a738088-8902-11e8-8e65-ed7e0bd56020.png">
<img width="754" alt="screen shot 2018-07-16 at 14 07 17" src="https://user-images.githubusercontent.com/10673976/42758016-4d2c6fb0-8902-11e8-8acd-2cbf3d9e6617.png">
<img width="376" alt="screen shot 2018-07-16 at 14 07 22" src="https://user-images.githubusercontent.com/10673976/42758019-50ec58ea-8902-11e8-80c9-9082c828a499.png">
<img width="412" alt="screen shot 2018-07-16 at 14 07 53" src="https://user-images.githubusercontent.com/10673976/42758036-5baf9c74-8902-11e8-8ab1-c6a0e3227da0.png">
<img width="774" alt="screen shot 2018-07-16 at 14 08 28" src="https://user-images.githubusercontent.com/10673976/42758043-60c0a348-8902-11e8-80a5-8a0e64c479da.png">
<img width="921" alt="screen shot 2018-07-16 at 14 09 00" src="https://user-images.githubusercontent.com/10673976/42758049-6618b0a6-8902-11e8-85ef-27aa6418b4df.png">

guillaumejenkins commented:
User experience looks good!

It would be cool down the road to support non-owners too (contributors), e.g. clone someone else's repo, make changes, push to a fork and open a PR back the the main repo.

guillaumejenkins commented on cli/server.ts:
```diff
@@ -99,41 +99,44 @@ function readAssetsAsync(logicalDirname: string): Promise<any> {
+        config: cfg,
+        files: []
+    };
+    
```
Linting error on this line: `ERROR: /home/travis/build/Microsoft/pxt/cli/server.ts[111, 1]: trailing whitespace`

mmoskal replied:
Right, fixed these, thanks!

guillaumejenkins commented:
Also, can we add a way to dynamically disable this? Something like URL query param, so that the Electron app can disable this

riknoll commented on pxtlib/sha1.ts:
```diff
@@ -0,0 +1,182 @@
+// Based on https://github.com/creationix/git-sha1
+// Converted to TypeScript and modified slightly
+
+namespace ts.pxtc.Util {
```
We will definitely need to get approval for this file. Not because of the license, but because it implements sha1

mmoskal replied:
@abchatra could you put in OSS req for this one?

The SHA-1 isn't security relevant here, but I'm not sure if that helps. Maybe there is some approved JS SHA-1 we could use?

abchatra commented:
This is pretty cool! I haven't looked at code yet. We discussed user experience with Jacqueline and one comment. 

_GitHub Login_ entry point should go to extensions page and not in the import. Login in import indicates user projects will be  saved in that login and it unintiutive that it is only for package writers. We need that space for office 365 login in future. 

Other than that it looks pretty cool.

mmoskal commented:
@guillaumejenkins does Electron disable all online interactions? If not, I see no reason to disable this. We would need a slightly different login experience (have the app long-poll on a server end point to get the token, which is obtained from the regular browser), but it's certainly doable.

As for the token - indeed our cloud sees it, but only in transit. We probably need to move the app to Microsoft github org.

As for the contributions, I guess the optimal solution would be to fork on import if the user has no write access to the repo, but PR creation is probably best handled by github web interface. I would leave this for later.

mmoskal commented:
@abchatra I moved the login link to "About" dialog. This is even more obscure than extensions dialogs, and we might keep it there as we refine the feature. I've also changed it to use personal access tokens for now.

However, I don't think this is only for package writers. I can easily imagine high school students collaborating on an Arcade project using GitHub - they would learn it in a quite soft way @eric-camplin 

Also, people can in fact use it for storing their own projects.

abchatra commented:
Students are not going to have GitHub login and we have plan for o365 this year. Package writers perspective, extensions page is better in my opinion. About is for versions and I am not sure login fits there. @Jaqster what is your opinion?

Is there any limitation in extensions page to add this?

Jaqster commented:
My vote is to put this feature in the Extensions dialog.  This seems like the most natural entry point for developers who want to write Extensions/Packages.  It's weird in the About page...  I would also say that this feature is most definitely for developers at this stage.  Yes, maybe in 2-3 years it will become more mainstream for students, but not today.  We can talk about how to surface this in the Extensions page... maybe a '+' card or something at the end of the list to prompt people to Add Your Own...

mmoskal commented:
I guess should be easy to add at the bottom of extensions page. At the top it's problematic since it's common modal stuff.

O365 is not going to cut the collaboration scenario, only storage. I think it might be also typical to teach source control together with a basic programming course. But let's just wait and see.

abchatra commented:
I registered crypto-js and it was auto approved https://ossmsft.visualstudio.com/_oss?requestId=659384&_a=progress

Please add it to 3rd party notices file

riknoll commented:
This needs to be behind a flag (the UI part). No need for this in the Cue app. Otherwise looks good!

mmoskal commented:
@riknoll @abchatra good to merge?

abchatra commented:
Yes go for it
